### PR TITLE
1-arg permutedims(df)

### PR DESCRIPTION
**Pull request by anandijain**
Going along with the other PR #3114, I was thinking maybe it would be better to just make permutedims easier to use instead of try to change the DataFrame constructors. 

The current methods for permutedims are honestly pretty confusing to me and I was surprised there wasn't a simple method that just permutes and doesn't add new columns or anything annoying like that.

I would love if someone could make it non-allocating or whatever, but I think regardless it makes sense to have this method defined. 

Would appreciate some comments :)

### CODE REVIEW

**Comment by bkamins** [comment]
Having thought about your proposal maybe indeed it makes sense.
We wanted to allow users to allow handling of old/new column names conveniently.
But, indeed, there are many cases, where one has a data frame of numbers and just wants what you propose.
Fortunately this would be consistent with the current design.

Let us wait for what other maintainers think.

CC @nalimilan

**Comment by nalimilan** [comment]
I guess it could make sense given that we already generate these `x1`... `xn` names in other places. In the `DataFrame` constructor you have to pass `:auto` explicitly for that, but not when using `=> AsTable` with `combine`.

The definition you propose destroys column names, but this has the advantage that if you have only numbers in the input the same is true of the output -- so it's closer to matrix transposition.

In what kind of use case does this arise?

**Comment by bkamins** [comment]
This PR is OK. It will need to be updated after https://github.com/JuliaData/DataFrames.jl/pull/3055 is merged. I will do the update of the code and of a docstring.

**Comment by bkamins** [comment]
@anandijain + @nalimilan : I have updated this PR in a way that I think is consistent and convenient. Please review (especially docstring updates - as usual 😄).

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -694,16 +702,40 @@ with name specified by `dest_namescol`.
...
+which always has element type `String`) will depend on the element types of
+_all_ input columns based on the result of `promote_type`.
 That is, if the source data frame contains `Int` and `Float64` columns,
 resulting columns will have element type `Float64`. If the source has
 `Int` and `String` columns, resulting columns will have element type `Any`.
 
 # Examples
 
 ```jldoctest
+julia> df
```
Comment (bkamins) [inline src/abstractdataframe/reshape.jl:715]:
```suggestion
julia> df = DataFrame(a=1:2, b=3:4)
```

File: test/reshape.jl [hunk 2]
```diff
@@ -668,6 +668,13 @@ end
     @test_throws ArgumentError permutedims(df, 1)
     # but allowed with strict=false
     @test permutedims(df, 1, strict=false) == ref
+
+    df = DataFrame(a=1:2, b=3:4)
+    @test permutedims(df) = DataFrame(a=[1, 3], b=[2, 4])
+    @test permutedims(df, [:p, :q]) = DataFrame(p=[1, 3], q=[2, 4])
+    @test permutedims(df, ["p", "q"]) = DataFrame(p=[1, 3], q=[2, 4])
+    @test_throws ArgumentError permutedims(df, ["p", "p"]) = DataFrame(p=[1, 3], q=[2, 4])
+    @test permutedims(df, ["p", "p"], makeunique=true) = DataFrame(p=[1, 3], p_1=[2, 4])
```
Comment (bkamins) [inline test/reshape.jl:677]:
```suggestion
    @test permutedims(df) == DataFrame(a=[1, 3], b=[2, 4])
    @test permutedims(df, [:p, :q]) == DataFrame(p=[1, 3], q=[2, 4])
    @test permutedims(df, ["p", "q"]) == DataFrame(p=[1, 3], q=[2, 4])
    @test_throws ArgumentError permutedims(df, ["p", "p"]) == DataFrame(p=[1, 3], q=[2, 4])
    @test permutedims(df, ["p", "p"], makeunique=true) == DataFrame(p=[1, 3], p_1=[2, 4])
```

File: test/reshape.jl [hunk 3]
```diff
@@ -668,6 +668,13 @@ end
     @test_throws ArgumentError permutedims(df, 1)
     # but allowed with strict=false
     @test permutedims(df, 1, strict=false) == ref
+
+    df = DataFrame(a=1:2, b=3:4)
+    @test permutedims(df) == DataFrame(a=[1, 3], b=[2, 4])
```
Comment (bkamins) [inline test/reshape.jl:673]:
```suggestion
    @test permutedims(df) == DataFrame(x1=[1, 3], x2=[2, 4])
```

**Comment by bkamins** [comment]
@anandijain - can you merge `main` into this PR. I need to add metadata handling to the added methods + correctness tests after you do it.

File: test/reshape.jl [hunk 4]
```diff
@@ -668,6 +668,13 @@ end
...
     # but allowed with strict=false
     @test permutedims(df, 1, strict=false) == ref
+
+    df = DataFrame(a=1:2, b=3:4)
+    @test permutedims(df) == DataFrame(x1=[1, 3], x2=[2, 4])
+    @test permutedims(df, [:p, :q]) == DataFrame(p=[1, 3], q=[2, 4])
+    @test permutedims(df, ["p", "q"]) == DataFrame(p=[1, 3], q=[2, 4])
+    @test_throws ArgumentError permutedims(df, ["p", "p"]) == DataFrame(p=[1, 3], q=[2, 4])
+    @test permutedims(df, ["p", "p"], makeunique=true) == DataFrame(p=[1, 3], p_1=[2, 4])
 end
```
Comment (nalimilan) [inline test/reshape.jl:678]: Also test what happens with empty data frames (both without and with columns)?

File: src/abstractdataframe/reshape.jl [hunk 5]
```diff
@@ -682,6 +683,13 @@ with name specified by `dest_namescol`.
 # Arguments
 - `df` : the `AbstractDataFrame`
 - `src_namescol` : the column that will become the new header.
+   Alternatively a vector of strings or `Symbol`s can be passed
+   which will be used as column names in returned data frame.
+   If this argument is not passed then auto-generated header is used with
+   column names `x1`, `x2`, ... and passing `makeunique` keyword argument is
+   not allowed.
+   If this argument is a vector or is skipped then then passing `dest_namescol`
+   and `strict` arguments is not allowed and column names from the `df` are dropped.
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl:692]: I wouldn't mention other arguments here, as it adds burden for users who may not care about them. Better say "Not supported when `src_namescol` is a vector or is omitted." when describing `makeunique`, `dest_namescol` and `strict`.
```suggestion
   If omitted then column names `x1`, `x2`, ... are generated automatically.
```

File: test/reshape.jl [hunk 6]
```diff
@@ -668,6 +668,13 @@ end
     @test_throws ArgumentError permutedims(df, 1)
     # but allowed with strict=false
     @test permutedims(df, 1, strict=false) == ref
+
+    df = DataFrame(a=1:2, b=3:4)
+    @test permutedims(df) == DataFrame(x1=[1, 3], x2=[2, 4])
+    @test permutedims(df, [:p, :q]) == DataFrame(p=[1, 3], q=[2, 4])
+    @test permutedims(df, ["p", "q"]) == DataFrame(p=[1, 3], q=[2, 4])
+    @test_throws ArgumentError permutedims(df, ["p", "p"]) == DataFrame(p=[1, 3], q=[2, 4])
+    @test permutedims(df, ["p", "p"], makeunique=true) == DataFrame(p=[1, 3], p_1=[2, 4])
```
Comment (bkamins) [inline test/reshape.jl:677]:
```suggestion
    @test permutedims(df, ["p", "p"], makeunique=true) == DataFrame(p=[1, 3], p_1=[2, 4])
    @test permutedims(DataFrame()) == permutedims(DataFrame(a=[], b=[])) ==
          permutedims(DataFrame(), []) == permutedims(DataFrame(a=[], b=[]), []) == DataFrame()
```

File: test/reshape.jl (hunk 4 above)
Comment (bkamins) [reply test/reshape.jl:678]: added

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
This PR cannot be merged yet. It must be merged with `main` and then I need to update metadata handling in it.

@anandijain - can you do the merge? Thank you!

**Comment by bkamins** [comment]
@nalimilan - are you OK that I merge this PR and then open next PR that updates this PR against metadata?

**Comment by nalimilan** [comment]
Sure.

File: src/abstractdataframe/reshape.jl [hunk 7]
```diff
@@ -682,28 +683,56 @@ with name specified by `dest_namescol`.
 # Arguments
 - `df` : the `AbstractDataFrame`
 - `src_namescol` : the column that will become the new header.
+   If omitted then column names `x1`, `x2`, ... are generated automatically.
```
Comment (bkamins) [inline src/abstractdataframe/reshape.jl:686]:
```suggestion
   If omitted then column names `:x1`, `:x2`, ... are generated automatically.
```

**Comment by bkamins** [comment]
Thank you!
